HW3/P5/label_regions.cl

```diff
@@ -61,39 +61,90 @@ propagate_labels(__global __read_write int *labels,
     if (idx_1D < buf_w) {
         for (int row = 0; row < buf_h; row++) {
-            buffer[row * buf_w + idx_1D] = 
-                get_clamped_value(labels,
-                                  w, h,
-                                  buf_corner_x + idx_1D, buf_corner_y + row);
+            buffer[row * buf_w + idx_1D] =
+            get_clamped_value(labels,
+                              w, h,
+                              buf_corner_x + idx_1D, buf_corner_y + row);
         }
     }
-
+    
     // Make sure all threads reach the next part after
     // the local buffer is loaded
     barrier(CLK_LOCAL_MEM_FENCE);
-
+    
     // Fetch the value from the buffer the corresponds to
     // the pixel for this thread
     old_label = buffer[buf_y * buf_w + buf_x];
-
+    
     // CODE FOR PARTS 2 and 4 HERE (part 4 will replace part 2)
     
+    //PART 2 CODES
+    //if(old_label < w * h){
+    //    buffer[buf_y * buf_w + buf_x] = labels[buffer[buf_y * buf_w + buf_x]];
+    //}
+    
+    //PART 4 CODES
+    
+    //Use the thread corresponding to the top-left corner as to execute the seriel implementation
+    if(lx == 0 && ly ==0){
+        int last_label = buffer[0];
+        int current_label;
+        
+        //row: local x index
+        for(uint row = 0; row < get_local_size(0); row++){
+            //col: local y index
+            for(uint col = 0; col < get_local_size(1); col++){
+                
+                //current_label_id: label index in buffer
+                int current_label_id = (row + halo) * buf_w + (col + halo);
+                current_label = buffer[current_label_id];
+                
+                //if foreground objects
+                if(current_label < w * h){
+                    if(current_label == last_label){
+                        //skip lookup, replace within the buffer
+                        buffer[current_label_id] = last_label;
+                    }else{
+                        //different from last fetch
+                        last_label = current_label;
+                        //reading global value.
+                        buffer[current_label_id] = labels[current_label];
+                    }
+                }
+            }
+        }
+    }
+    
+    barrier(CLK_LOCAL_MEM_FENCE);
+    
     // stay in bounds
     if ((x < w) && (y < h)) {
         // CODE FOR PART 1 HERE
         // We set new_label to the value of old_label, but you will need
         // to adjust this for correctness.
+        
         new_label = old_label;
-
+        
+        if(new_label < w * h){
+            int up_label = buffer[(buf_y - 1) * buf_w + buf_x];
+            int down_label = buffer[(buf_y + 1) * buf_w + buf_x];
+            int left_label = buffer[buf_y * buf_w + buf_x - 1];
+            int right_label = buffer[buf_y * buf_w + buf_x + 1];
+            
+            new_label = min(min(min(min(old_label, up_label), down_label), left_label), right_label);
```

Comment:
After parts 2 and 4, you should use buffer[buf_w * buf_y + buf_x] instead of old_label.